src/DefaultBuilder/src/WebHost.cs

```diff
@@ -244,9 +244,15 @@ internal static void ConfigureWebDefaultsSlim(IWebHostBuilder builder)
     internal static void ConfigureWebDefaultsEmpty(IWebHostBuilder builder)
     {
         // Note this doesn't configure any server - Kestrel or otherwise.
-        // It is "empty" and up to the caller to configure a server.
+        // It also doesn't register HostFiltering or ForwardedHeaders.
+        // It is "empty" and up to the caller to configure these services.
 
-        ConfigureWebDefaultsWorker(builder, configureRouting: null);
+        builder.ConfigureServices((hostingContext, services) =>
+        {
+            // Routing services need to be registered in order for WebApplication to
+            // implement IEndpointRouteBuilder (i.e. MapGet, etc) correctly.
+            services.AddRoutingCore();
```

Comment:
Why even call AddRoutingCore?

Reply:
'cos then `app.MapGet` methods and friends don't work. Routing is a core requirement in the `WebApplication` model.

Reply:
Also for the simple reasoning of "because @halter73 told me to".

Reply:
The one drawback I can think of having this is that AddRoutingCore does bring in app size that may not be required.

If I publish the following app for NativeAOT:

```C#
var builder = WebApplication.CreateEmptyBuilder(new ());
builder.WebHost.UseKestrelCore();

var app = builder.Build();

app.Use(async (context, next) =>
{
    await context.Response.WriteAsync("Hello, World!");
    await next(context);
});

app.Run();
```

on win-x64 I get:

* 7.98 MB with AddRoutingCore
* 7.50 MB without

So AddRoutingCore adds roughly 500 KB of app size, even when it isn't used.

Reply:
How does it fail if you try to call `MapGet` without adding the routing services & middleware?

Reply:
CreateEmptyBuilder doesn’t need to be usable by default. It just needs to be empty. We can make sure the errors are good but empty should really mean empty.

Reply:
> How does it fail if you try to call `MapGet` without adding the routing services & middleware?

You get the following exception at startup:

```
Unhandled exception. System.InvalidOperationException: Unable to find the required services. Please add all the required services by calling 'IServiceCollection.AddRouting' inside the call to 'ConfigureServices(...)' in the application startup code.
```

If we were to not register the routing services by default, you'd start off with the following code:

```csharp
var builder = WebApplication.CreateEmptyBuilder(new());
var app = builder.Build();
app.MapGet("/", () => "Hello world!");
app.Run();
```

You'd hit an exception about no `IServer` implementation being registered which you'd fix with this:

```diff
var builder = WebApplication.CreateEmptyBuilder(new());
+builder.WebHost.UseKestrel();
var app = builder.Build();
app.MapGet("/", () => "Hello world!");
app.Run();
```

Then you'd get the exception about the lack of routing services which you'd fix with this:

```diff
var builder = WebApplication.CreateEmptyBuilder(new());
builder.WebHost.UseKestrel();
+builder.Services.AddRouting();
var app = builder.Build();
app.MapGet("/", () => "Hello world!");
app.Run();
```

Before you have a functioning app.

> We can make sure the errors are good but empty should really mean empty.

There's a chance we can improve the experience you get here by producing a single exception that outlines the full requirements for WebApplication instead of two.

Although I hear the argument about routing services being required to fulfill the spirit of the `WebApplication` (as in it implement `IEndpointBuilder`), it feels like if we're going through the effort to make a _truly_ empty builder, then we shouldn't compromise on that principal.

Reply:
One other thing that struck me weird about this is the fact that everywhere else in the framework (AFAIK) we register middleware in response to services existing in DI. This is the first example I've seen where we eagerly register services in response to anticipating we'll add a certain middleware.

Reply:
Routing shouldn't be required for this right (ignoring current implementation details for a moment):

```csharp
var builder = WebApplication.CreateEmptyBuilder();
var app = builder.Build();
app.Use(async (context, next) => await next(context)); 
app.Run();
```

Reply:
Assume ```Use(...)``` does something interesting and is the end of the line :)

Reply:
> Routing shouldn't be required for this right (ignoring current implementation details for a moment):

No, you don't need routing services for that case.

Reply:
I was going to suggest adding the AddRoutingCore services conditionally on a EndpointDataSource being added to the WebApplication similar to what we do for the middleware, but then realized the service provider has already been built.

It'd be nice if we could make it "just work", but that doesn't seem easy. I'm okay with leaving out AddRoutingCore as long as the error is clear enough which it appears to be. Maybe we should update the exception message to recommend AddRouting or AddRoutingCore.

Reply:
The other thing to keep in mind is that we don't need to really worry about whether this would work out of the box because it's unlikely that we'll ever have this in a template where that is a requirement.

We tend to analyse these decisions as if looking at what the minimal working example would be (generally a good idea). But that isn't a requirement here.

Reply:
I've removed AddRoutingCore from the empty builder.

I didn't update the exception message because it is used in a few places and tries to be generic. Updating it to include "or AddRoutingCore" would have made it be not generic anymore. So I left it. We can do it in a separate PR if required.